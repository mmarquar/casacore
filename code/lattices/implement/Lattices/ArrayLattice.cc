//# ArrayLattice.cc: this defines the Lattice wrapper class for Arrays.
//# Copyright (C) 1995,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <trial/Lattices/ArrayLattice.h>
#include <trial/Lattices/ArrLatticeIter.h>
#include <trial/Lattices/LatticeIterInterface.h>
#include <trial/Lattices/LatticeNavigator.h>
#include <aips/Exceptions/Error.h>
#include <aips/Lattices/IPosition.h>
#include <aips/Lattices/Slicer.h>
#include <aips/Utilities/Assert.h>
#include <aips/Utilities/COWPtr.h>

template<class T>
ArrayLattice<T>::ArrayLattice()
: itsWritable (False)
{}

template<class T>
ArrayLattice<T>::ArrayLattice (const IPosition& shape) 
: itsData     (shape),
  itsWritable (True)
{}

template<class T>
ArrayLattice<T>::ArrayLattice (Array<T>& array) 
: itsData     (array),
  itsWritable (True)
{}

template<class T>
ArrayLattice<T>::ArrayLattice (const Array<T>& array) 
: itsData     (array),
  itsWritable (False)
{}

template<class T>
ArrayLattice<T>::ArrayLattice (const ArrayLattice<T>&other) 
: itsData     (other.itsData),
  itsWritable (other.itsWritable)
{
    // Nothing
}

template<class T>
ArrayLattice<T>::~ArrayLattice()
{
  // Nothing
}

template<class T>
ArrayLattice<T>& ArrayLattice<T>::operator= (const ArrayLattice<T>& other)
{
  if (this != &other) {
    itsData     = other.itsData;
    itsWritable = other.itsWritable;
  }
  return *this;
}

template<class T>
Lattice<T>* ArrayLattice<T>::clone() const
{
  return new ArrayLattice<T> (*this);
}

template <class T>
Bool ArrayLattice<T>::isWritable() const
{
  return itsWritable;
}

template<class T>
IPosition ArrayLattice<T>::shape() const
{
  return itsData.shape();
} 

template<class T>
Bool ArrayLattice<T>::getSlice (COWPtr<Array<T> >& bufPtr,
				const IPosition& start, 
				const IPosition& shape,
				const IPosition& stride,
				Bool removeDegenerateAxes) const
{
  return getSlice (bufPtr, Slicer(start, shape, stride), removeDegenerateAxes);
}

template<class T>
Bool ArrayLattice<T>::getSlice (COWPtr<Array<T> >& bufPtr,
				const Slicer& section, 
				Bool removeDegenerateAxes) const
{
  // cast away the constness of the ArrayLattice using a pointer copy. This can
  // be done as the COWPtr will be set to be "readonly" and hence the
  // ArrayLattice cannot be modified without the COWPtr making a copy of the
  // cursor
  // The COWPtr takes over the pointer to the array.
  ArrayLattice<T>* This = (ArrayLattice<T>*) this;
  Array<T>* arr = new Array<T>;
  Bool isARef = This->getSlice(*arr, section, removeDegenerateAxes);
  bufPtr = COWPtr<Array<T> > (arr, True, isARef);
  // While the returned array is normally a reference return "False" indicating
  // a copy as any attempt to modify the Array will result in a copy.
  return False;
}

template<class T>
Bool ArrayLattice<T>::getSlice (Array<T>& buffer,
				const IPosition& start,
				const IPosition& shape, 
				const IPosition& stride,
				Bool removeDegenerateAxes)
{
  return getSlice (buffer, Slicer(start, shape, stride), removeDegenerateAxes);
}

template<class T>
Bool ArrayLattice<T>::getSlice (Array<T>& buffer,
				const Slicer& section, 
				Bool removeDegenerateAxes)
{
  // The following block checks that the supplied buffer is the right size or
  // empty. These restrictions are not required for the rest of this function
  // to work. I impose them because they are required by the corresponding
  // functions in the PagedArray class. If these restrictions are a performance
  // bottleneck they can by removed (only for optimised code please).
  if (buffer.nelements() != 0) {
    const IPosition& shape = buffer.shape();
    if (removeDegenerateAxes) {
      AlwaysAssert (shape.isEqual (section.length().nonDegenerate()),
		    AipsError);
    } else {
      AlwaysAssert (shape.isEqual (section.length()), AipsError);
    }
  }
  Array<T> cursor = itsData(section.start(), section.end(), section.stride());
  if (removeDegenerateAxes) {
    buffer.nonDegenerate(cursor);
  } else {
    buffer.reference(cursor);
  }
  return True;
}

template<class T>
void ArrayLattice<T>::putSlice (const Array<T>& sourceBuffer,
				const IPosition& where, 
				const IPosition& stride)
{
  if (!itsWritable) {
      throw (AipsError ("ArrayLattice::putSlice - non-writable lattice"));
  }
  const uInt sdim = sourceBuffer.ndim();
  const uInt ldim = ndim();
  DebugAssert(ldim == where.nelements(), AipsError);
  DebugAssert(ldim == stride.nelements(), AipsError);
  if (sdim == ldim) {
    itsData(where, 
	    where + (sourceBuffer.shape()-1)*stride, 
	    stride) = sourceBuffer;
  } else {
    Array<T> allAxes(sourceBuffer.addDegenerate(ldim-sdim));
    itsData(where, 
	    where + (allAxes.shape()-1)*stride, 
	    stride) = allAxes;
  }
}

template<class T>
void ArrayLattice<T>::putSlice (const Array<T>& sourceBuffer,
				const IPosition& where)
{
  if (!itsWritable) {
      throw (AipsError ("ArrayLattice::putSlice - non-writable lattice"));
  }
  const uInt sdim = sourceBuffer.ndim();
  const uInt ldim = ndim();
  DebugAssert(ldim == where.nelements(), AipsError);
  if (sdim == ldim) {
    itsData(where, 
	    where + (sourceBuffer.shape()-1)) = sourceBuffer;
  } else {
    Array<T> allAxes(sourceBuffer.addDegenerate(ldim-sdim));
    itsData(where, 
	    where + (allAxes.shape()-1)) = allAxes;
  }
}

template<class T>
void ArrayLattice<T>::getIterSlice (Array<T>& buffer, const IPosition& start,
				    const IPosition& end, const IPosition& incr)
{
    Array<T> tmp (itsData(start, end, incr));
    buffer.reference (tmp);
}


template<class T>
void ArrayLattice<T>::set (const T& value)
{
  if (!itsWritable) {
      throw (AipsError ("ArrayLattice::set - non-writable lattice"));
  }
  itsData.set(value);
}

template<class T>
T ArrayLattice<T>::getAt (const IPosition& where) const
{
  return itsData(where);
}

template<class T>
void ArrayLattice<T>::putAt (const T& value, const IPosition& where)
{
  if (!itsWritable) {
      throw (AipsError ("ArrayLattice::putAt - non-writable lattice"));
  }
  itsData(where) = value;
}

template<class T>
LatticeIterInterface<T>* ArrayLattice<T>::makeIter
                                       (const LatticeNavigator& nav) const
{
  return new ArrLatticeIter<T>(*this, nav);
}

template<class T>
Array<T>& ArrayLattice<T>::asArray()
{
  if (!itsWritable) {
      throw (AipsError ("ArrayLattice::asArray - non-writable lattice"));
  }
  return itsData;
}

template<class T>
const Array<T>& ArrayLattice<T>::asArray() const
{
  return itsData;
}

// Check class invariants. 
template<class T>
Bool ArrayLattice<T>::ok() const
{
  return itsData.ok();
}
