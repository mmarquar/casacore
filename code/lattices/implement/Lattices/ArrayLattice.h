//# ArrayLattice: Object which converts an Array to a Lattice.
//# Copyright (C) 1994,1995,1996,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#if !defined(AIPS_ARRAYLATTICE_H)
#define AIPS_ARRAYLATTICE_H

#include <aips/aips.h>
#include <trial/Lattices/Lattice.h>
#include <aips/Arrays/Array.h>

class IPosition;
class LatticeNavigator;
class Slicer;
template <class T> class COWPtr;
template <class T> class RO_LatticeIterInterface;
template <class T> class LatticeIterInterface;

// <summary>
// A memory resident Lattice
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="tArrayLattice" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Lattice>Lattice</linkto>
//   <li> <linkto class=Array>Array</linkto>
// </prerequisite>

// <etymology>
// The ArrayLattice name reflects its role as a Lattice interface to an Array
// object.
// </etymology>

// <synopsis> 
// An ArrayLattice is a concrete Lattice class were the data is stored in
// memory as opposed to the <linkto class=PagedArray>PagedArray</linkto> class
// where the data is stored on disk. As a result this class is much more
// suitable to problems which require small Lattices that can fit into the
// memory of a computer. 
//
// ArrayLattices impose another layer of function calls on top of a an
// Array. As a result they should not be used for generic Array
// manipulation. They are useful if you have an Array that needs to use
// Lattice functions or exchange data with PagedArrays or other Lattice
// derivatives (if they ever get written). For example the LatticeIterator
// class can iterate through an Array in more ways than any of the
// ArrayIterator classes can. The examples below illustrate some uses for
// ArrayLattices. 
// </synopsis> 

// <example>
// All the examples in this section are available in
// <src>dArrayLattice.cc</src>
//
// <h4>Example 1:</h4>
// In this example an Array of data is converted into an ArrayLattice so that
// the CopyLattice function can be used to write the data to a PagedArray which
// will be stored on disk. This may be conceptually simpler than having to use
// the putSlice member function in the PagedArray class and is only slightly
// less efficient.
// <srcblock>
// // make an Array and fill it with data.
// Array<Float> myArray(IPosition(3, 64, 64, 2));
// indgen(myArray); // fills the Array with 0,1,2,....,64*64*2-1
// // construct the ArrayLattice
// ArrayLattice<Float> myLattice(myArray);
// // make a PagedArray to store the data on disk
// PagedArray<Float> myPagedArray(myLattice.shape(), "myTestData.array");
// // now copy the data onto disk
// CopyLattice(myPagedArray, myLattice);
// </srcblock>
//
// <h4>Example 2:</h4>
// The <linkto class=ArrayIterator>ArrayIterator</linkto> class (or its
// derivatives the <linkto class=VectorIterator>VectorIterator</linkto> and the
// <linkto class=MatrixIterator>MatrixIterator</linkto> classes) do not allow
// the user to specify a cursor shape. In this example a Cube class will be
// converted into an ArrayLattice so that an ArrLatticeIter can be used to
// access the data spectrum by spectrum (assuming the z-axis is frequency).
//
// <srcblock>
// Cube<Float> arr(64,64,128);
// // assume that the data gets put into the cube somehow
// // now construct an ArrayLattice from this cube.
// ArrayLattice<Float> lat(arr);
// // Construct an iterator that returns the 128-element spectra one at a time
// ArrLatticeIter<Float> iter(lat, IPosition(3,1,1,128));
// // construct a Matrix to hold the results
// Matrix<Float> channelSum(64,64);
// // and do the summation one spectrum at a time
// for (iter.reset(); !iter.atEnd(); iter++)
//    channelSum(iter.position().getFirst(2)) = sum(iter.cursor());
// </srcblock>
//
//  There are more examples in the <linkto class=Lattice>Lattice</linkto> class
//  and many of the examples in the 
// <linkto class=PagedArray>PagedArray</linkto> class will also be instructive.
// </example>

// <motivation>
// We needed a way of creating Lattices but with AIPS++ Array characteristics.
// </motivation>

//# <todo asof="1997/05/31">
//# </todo>

// <linkfrom anchor="ArrayLattice" classes="Lattice PagedArray">
//  <here>ArrayLattice</here> - a memory based Lattice.
// </linkfrom>


template<class T>
class ArrayLattice : public Lattice<T>
{
public: 
  // The default constructor creates a ArrayLattice that is useless for just
  // about everything, except that it can be assigned to with the assignment
  // operator.
  ArrayLattice();

  // Construct an ArrayLattice with the specified shape.
  // This results in a writable lattice.
  ArrayLattice (const IPosition& shape);

  // Construct an ArrayLattice that references the given Array.
  // This results in a writable lattice.
  ArrayLattice (Array<T>& array);

  // Construct an ArrayLattice that references the given Array.
  // This results in a non-writable lattice.
  ArrayLattice (const Array<T>& array);

  // the copy constructor which uses reference semantics.
  ArrayLattice (const ArrayLattice& other);

  // the destructor does very little
  ~ArrayLattice();

  // the assignment operator which uses copy semantics.
  ArrayLattice& operator= (const ArrayLattice& other);

  // Make a copy of the object (reference semantics).
  virtual Lattice<T>* clone() const;

  // Is the lattice writable?
  virtual Bool isWritable() const;

  // returns the shape of the ArrayLattice.
  virtual IPosition shape() const; 
  
  // Functions which extract an Array of values from the ArrayLattice. The
  // arguments to these functions are:
  // <ul>
  // <li> buffer: an <src>Array<T></src> or a
  // <src>COWPtr<Array<T>></src>. The buffer shape must be big enough to
  // exactly contain the specified slice. Otherwise an exception will be
  // thrown. Alternatively the buffer can be empty
  // (buffer.nelements()==0). Then the buffer will be resized to fit the
  // requested slice. Unlike the other arguments the buffer can have fewer
  // axes than the underlying PagedArray, but then you must set the
  // removeDegenerateAxes flag to True.
  // <li> start: The starting position (or Bottom Left Corner), within the
  // PagedArray, of the slice to be extracted. Must have the same number of
  // axes as the PagedArray otherwise an exception will be thrown.
  // <li> shape: The shape of the slice to be extracted. This is not a
  // position within the PagedArray but the actual shape the buffer will
  // have after this function is called althoug it always includes all the
  // degenerate axes even if the buffer does not. This argument added to the
  // "start" argument will be the "Top Right Corner", assuming the stride is
  // one on all axes.
  // <li> stride: The increment for each axis. A stride of one will return
  // every data element, a stride of two will return every other
  // element. The IPosition elements may be different for each respective
  // axis. Thus, a stride of IPosition(2,1,2) says: fill the buffer with
  // every element whose position has a first index between start(0) and
  // start(0)+shape(0) and a second index which is every other element
  // between start(1) and (start(1)+shape(1)*2). Must have the same number
  // of axes as the PagedArray otherwise an exception will be thrown.
  // <li> section: The preferred way of specifying the start, shape and
  // stride.
  // <li> removeDegenerateAxes: a Bool which indicates whether to remove axes
  // of length one in the extracted slice. You must set this to True if the
  // supplied buffer has fewer axes than the specified slice, otherwise an
  // exception is thrown. If the supplied buffer is empty all degenerate
  // axes are stripped from the returned buffer.
  // </ul>
  // These functions return False if the returned buffer is a copy of the
  // actual data and True if the returned buffer is a reference to the data in
  // the ArrayLattice. You do not need to call putSlice to put the buffer back
  // into the ArrayLattice if the buffer is a reference to the data. Currently
  // the const versions of getSlice return False and the non-const versions
  // (ie. the ones with <src>Array<T></src> and not <src>COWPtr<Array<T>></src>
  // arguments) return True.
  // <group>  
  virtual Bool getSlice (COWPtr<Array<T> >& bufPtr, const IPosition& start, 
			 const IPosition& shape, const IPosition& stride,
			 Bool removeDegenerateAxes=False) const;
  virtual Bool getSlice (COWPtr<Array<T> >& bufPtr, const Slicer& section, 
			 Bool removeDegenerateAxes=False) const;
  virtual Bool getSlice (Array<T>& buffer, const IPosition& start, 
			 const IPosition& shape, const IPosition& stride,
			 Bool removeDegenerateAxes=False);
  virtual Bool getSlice (Array<T>& buffer, const Slicer& section, 
			 Bool removeDegenerateAxes=False);
  // </group>

  // A function which copies an Array of values into the ArrayLattice at the
  // location specified by the IPosition "where", incrementing by
  // "stride". All of the IPosition arguments must have the same number of
  // dimensions as the ArrayLattice. The sourceBuffer array may have fewer
  // axes than the ArrayLattice. The stride defaults to one if not specified.
  // <group>
  virtual void putSlice (const Array <T>& sourceBuffer, 
			 const IPosition& where, const IPosition& stride);
  virtual void putSlice (const Array <T>& sourceBuffer, 
			 const IPosition& where);
  // </group>
  
  // functions which sets all of the elements in the Lattice to a value.
  virtual void set (const T& value);

  // functions which returns an Array of the data within this Lattice.
  // <group>
  Array<T>& asArray();
  const Array<T>& asArray() const;
  // </group>

  // These are the true implementations of the parenthesis operator. It will
  // probably be more convienient to use the actual parenthesis operator
  // defined in the Lattice base class.
  // <group>
  virtual T getAt (const IPosition& where) const;
  virtual void putAt (const T& value, const IPosition& where);
  // </group>
  
  // a function which checks for internal consistency. Returns False if
  // something nasty has happened to the ArrayLattice.
  virtual Bool ok() const;
  
  // This function is used by the LatticeIterator class to generate an
  // iterator of the correct type for a specified Lattice. Not recommended
  // for general use. 
  virtual LatticeIterInterface<T>* makeIter(
				   const LatticeNavigator& navigator) const;

  // Get a slice in an optimized way (specifically for ArrLatticeIter).
  // It returns in <src>buffer</src> a reference to the lattice array.
  void getIterSlice (Array<T>& buffer, const IPosition& start,
		     const IPosition& end, const IPosition& incr);

private:
  Array<T> itsData;
  Bool     itsWritable;
};


#endif
