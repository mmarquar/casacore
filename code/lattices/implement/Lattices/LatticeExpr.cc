//# LatticeExpr.cc:  this defines LatticeExpr.cc
//# Copyright (C) 1997,1998,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <trial/Lattices/LatticeExpr.h>
#include <trial/Lattices/LELArray.h>
#include <aips/Arrays/Array.h>
#include <aips/Lattices/Slicer.h>
#include <aips/Utilities/Assert.h>
#include <aips/Exceptions/Error.h> 

typedef LELArray<Bool> gppbug_latticeexpr_bool;
typedef LELArray<Float> gppbug_latticeexpr_float;
typedef LELArray<Double> gppbug_latticeexpr_double;
typedef LELArray<Complex> gppbug_latticeexpr_complex;
typedef LELArray<DComplex> gppbug_latticeexpr_dcomplex;


template <class T>
LatticeExpr<T>::LatticeExpr()
: lastChunkPtr_p (0)
{}

template <class T>
LatticeExpr<T>::LatticeExpr (const LatticeExprNode& expr, uInt)
: lastChunkPtr_p (0)
//
// Construct from a LatticeExprNode object.  The LEN type is
// converted to match the template type if possible
//
{
    DataType thisDT = whatType ((T*)0);
    if (expr.dataType() == thisDT) {
	expr_p = expr;
    } else {
	if (expr.dataType() != TpBool) {
	    throw (AipsError ("LatticeExpr::constructor - "
			      "Bool cannot be converted to a numeric type"));
	}
	switch (thisDT) {
	case TpFloat:
	    expr_p = expr.makeFloat();
	    break;
	case TpDouble:
	    expr_p = expr.makeDouble();
	    break;
	case TpComplex:
	    expr_p = expr.makeComplex();
	    break;
	case TpDComplex:
	    expr_p = expr.makeDComplex();
	    break;
	default:
	    throw (AipsError ("LatticeExpr::constructor - "
			      "A numeric type cannot be converted to Bool"));
	}
    }
}

template <class T>
LatticeExpr<T>::~LatticeExpr()
{
   delete lastChunkPtr_p;
}

template <class T>
LatticeExpr<T>::LatticeExpr (const LatticeExpr<T>& other)
: expr_p (other.expr_p),
  lastChunkPtr_p (0)
//
// Copy constructor.  Uses reference semantics
//
{}

template <class T>
LatticeExpr<T>& LatticeExpr<T>::operator=(const LatticeExpr<T>& other)
//
// Assignment. Uses reference semantics
//
{
   if (this != &other) {
      expr_p = other.expr_p;
      delete lastChunkPtr_p;
      lastChunkPtr_p = 0;
      lastSlicer_p = Slicer();
   }
   return *this;
}


template <class T>
Lattice<T>* LatticeExpr<T>::clone() const
//
// Return a copy of the LatticeExpr object. Uses
// reference semantics.
{
   return new LatticeExpr (*this);
}

template <class T>
MaskedLattice<T>* LatticeExpr<T>::cloneML() const
//
// Return a copy of the LatticeExpr object. Uses
// reference semantics.
{
   return new LatticeExpr (*this);
}

template <class T>
Bool LatticeExpr<T>::isMasked() const
{
   return expr_p.isMasked();
}

template <class T>
const LatticeRegion& LatticeExpr<T>::region() const
{
   return region_p;
}

template <class T>
Bool LatticeExpr<T>::isWritable() const
//
// A LatticeExpr lattice is not writable
//
{
   return False;
}

template <class T>
IPosition LatticeExpr<T>::shape() const
{
   return expr_p.shape();
}
  
template <class T>
IPosition LatticeExpr<T>::doNiceCursorShape (uInt) const
{
   return expr_p.getAttribute().tileShape();
}
  
template <class T>
LatticeCoordinates LatticeExpr<T>::coordinates() const
{
   return expr_p.getAttribute().coordinates();
}
  
template <class T>
Bool LatticeExpr<T>::doGetSlice (Array<T>& buffer,
				 const Slicer& section)
{
// Evaluate the expression if not accessing the same section again.
   if (!(section.start().isEqual (lastSlicer_p.start())
     &&  section.end().isEqual (lastSlicer_p.end())
     &&  section.stride().isEqual (lastSlicer_p.stride()))) {
      delete lastChunkPtr_p;
      lastChunkPtr_p = new LELArray<T> (section.length());
      lastSlicer_p = section;
      expr_p.eval (*lastChunkPtr_p, section);
   }
   buffer.reference (lastChunkPtr_p->value());
   return True;
}

template <class T>
Bool LatticeExpr<T>::doGetMaskSlice (Array<Bool>& buffer,
				     const Slicer& section)
{
// Evaluate if masked and if different section.
   if (expr_p.isMasked()) {
      if (!(section.start().isEqual (lastSlicer_p.start())
        &&  section.end().isEqual (lastSlicer_p.end())
        &&  section.stride().isEqual (lastSlicer_p.stride()))) {
	 delete lastChunkPtr_p;
	 lastChunkPtr_p = new LELArray<T> (section.length());
	 lastSlicer_p = section;
	 expr_p.eval (*lastChunkPtr_p, section);
      }
      if (lastChunkPtr_p->isMasked()) {
	 buffer.reference (lastChunkPtr_p->mask());
	 return True;
      }
   }
// Not masked, so we can simply fill the buffer with True values.
   buffer.resize (section.length());
   buffer = True;
   return False;
}

template <class T>
void LatticeExpr<T>::doPutSlice (const Array<T>&, const IPosition&,
				 const IPosition&)
{
   throw (AipsError ("LatticeExpr::putSlice - is not possible"));
}

template<class T>
void LatticeExpr<T>::copyDataTo (Lattice<T>& to) const
{
// If a scalar, set lattice to its value.
// Otherwise use the Lattice copyDataTo function.

  if (expr_p.isScalar()) {
    T value;
    expr_p.eval (value);
    to.set (value);
  } else {
    Lattice<T>::copyDataTo (to);
  }
}
