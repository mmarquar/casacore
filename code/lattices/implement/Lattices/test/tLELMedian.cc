//# tLELMedian.cc:  Tests the median function in LELFunction
//# Copyright (C) 1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <trial/Lattices/ArrayLattice.h>
#include <trial/Lattices/TempLattice.h>
#include <trial/Lattices/LatticeIterator.h>
#include <trial/Lattices/LELFunction.h>
#include <trial/Lattices/LatticeExpr.h>
#include <trial/Lattices/LELScalar.h>
#include <aips/Arrays/Array.h>
#include <aips/Arrays/ArrayMath.h>
#include <aips/Exceptions/Error.h>
#include <aips/Inputs/Input.h>
#include <aips/Lattices/IPosition.h>
#include <aips/OS/Timer.cc>
#include <iostream.h>


main (int argc, char *argv[])
{
  try {
    cout << ">>>" << endl;
    Input inp(1);
    inp.version(" ");
    inp.create("nx", "16", "Number of pixels along the x-axis", "int");
    inp.create("ny", "16", "Number of pixels along the y-axis", "int");
    inp.readArguments(argc, argv);
    cout << "<<<" << endl;

    const uInt nx=inp.getInt("nx");
    const uInt ny=inp.getInt("ny");

//
// Test various sized arrays.
//

    {
      IPosition shape(2,nx,ny);
      Array<Float> arr(shape);
      indgen (arr);
      ArrayLattice<Float> aF(arr);

      cout << median (arr) << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 16).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 4).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 2).value() << endl;
      LatticeExprNode afExpr(aF);
      LatticeExpr<Float> expr(afExpr[aF>4]);
      cout << LELFunctionReal1D<Float>::maskedMedian(expr).value() << endl;
    }

    {
      IPosition shape(2,10*nx,10*ny);
      Array<Float> arr(shape);
      indgen (arr);
      ArrayLattice<Float> aF(arr);

      cout << median (arr) << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 16).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 4).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 2).value() << endl;
      LatticeExprNode afExpr(aF);
      LatticeExpr<Float> expr(afExpr[aF>4]);
      cout << LELFunctionReal1D<Float>::maskedMedian(expr).value() << endl;
    }

    {
      IPosition shape(2,32*nx,32*ny);
      Array<Float> arr(shape);
      indgen (arr);
      ArrayLattice<Float> aF(arr);

      cout << median (arr) << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 128).value() << endl;
      LatticeExprNode afExpr(aF);
      LatticeExpr<Float> expr(afExpr[aF>4]);
      cout << LELFunctionReal1D<Float>::maskedMedian(expr).value() << endl;
    }

    // Hereafter numbers can be different on different machines.
    // So 'outcomment' it for assay (and also outcomment the timings).
    cout << ">>>" << endl;
    {
      IPosition shape(2,100*nx,100*ny);
      Array<Float> arr(shape);
      indgen (arr, float(0), float(0.01));
      ArrayLattice<Float> aF(arr);
      cout << "Last value = " << arr(shape-1) << endl;
      Timer timer;

      cout << median (arr) << endl;
      timer.show ("normal median");
      timer.mark();
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      timer.show ("ArrayLattice ");
      timer.mark();
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 1280).value()
 << endl;
      timer.show ("ArrayLat 1280");

      timer.mark();
      LatticeExprNode afExpr(aF);
      LatticeExpr<Float> expr(afExpr[aF>4]);
      cout << LELFunctionReal1D<Float>::maskedMedian(expr).value() << endl;
      timer.show ("MaskedLattice");
    }

    {
      IPosition shape(2,100*nx,100*ny);
      Array<Float> arr(shape);
      indgen (arr, float(0), float(0.01));
      TempLattice<Float> aF(shape);
      aF.put (arr);
      Timer timer;

      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      timer.show ("PagedArray   ");
      timer.mark();
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 4).value() << endl;
      timer.show ("PagedArr 4   ");
      timer.mark();
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 1280).value()
 << endl;
      timer.show ("PagedArr 1280");
    }

    {
      IPosition fshape(2,500*nx,500*ny);
      TempLattice<Float> aF(fshape);
      IPosition shape = aF.niceCursorShape();
      cout << "tileshape = " << shape << endl;
      Array<Float> arr(shape);
      indgen (arr, float(0), float(0.01));
      Timer timer;
      LatticeIterator<Float> iter(aF, shape);
      while (! iter.atEnd()) {
	iter.woCursor() = arr;
	iter++;
      }
      timer.show ("Fill PagedArr");
      timer.mark();

      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF).value() << endl;
      timer.show ("BigPagedArray");
      timer.mark();
      cout << LELFunctionReal1D<Float>::unmaskedMedian(aF, 1280*25).value()
	   << endl;
      timer.show ("PagedArr 1280");
    }
    cout << "<<<" << endl;

  } catch (AipsError& x) {
    cout << "Unexpected exception: " << x.getMesg() << endl;
    exit(1);
  }

  exit(0);
}
