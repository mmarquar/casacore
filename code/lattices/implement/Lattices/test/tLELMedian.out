>>>
/home/rzmws0/gvd/aips++/sun4sol_gnu/bin/tLELMedian: Version  
<<<
127
127
127
127
127
130
12799
12799
12799
12799
12799
12802
131071
131071
131071
131074
>>>
Last value = 25099.6
12599.6
normal median        0.3 real        0.24 user        0.06 system
12599.6
ArrayLattice        1.03 real        1.02 user           0 system
12599.6
ArrayLat 1280       1.05 real        1.03 user        0.01 system
12601.6
MaskedLattice       2.73 real        2.14 user        0.55 system
12599.6
PagedArray          1.04 real        1.02 user        0.01 system
12599.6
PagedArr 4          1.29 real        1.29 user           0 system
12599.6
PagedArr 1280       1.03 real        1.03 user           0 system
tileshape = [200, 160]
Fill PagedArr      41.01 real         6.4 user        7.56 system
159.981
BigPagedArray       80.6 real       19.94 user        7.58 system
159.981
PagedArr 1280     155.36 real       26.27 user       21.22 system
<<<
